File: lore/mapas/ferramentas/static/js/lugares.js
// Ferramenta de Lugar (ponto) -- B2 (2026-09-22), reescrita em 2026-09-23 com
// modal próprio no lugar dos prompt()/confirm() nativos, símbolo por tipo,
// tamanho por importância e lista lateral com busca.
//
// Regra de confirmação (item 3i do pedido): APAGAR UM LUGAR NÃO PERGUNTA, porque
// o desfazer cobre (é uma operação registrada em operacoes.py). Confirmação fica
// só pra ação destrutiva SEM desfazer -- hoje, só "limpar todas as medições" da
// régua, que é estado do navegador e não passa pelo log de operações.

const TIPOS_LUGAR = ["cidade", "vila", "fortaleza", "porto", "ruina", "marco"];
const IMPORTANCIAS_LUGAR = ["pequena", "media", "grande"]; // null = não classificado (mesmo tratamento visual de "pequena")

// Símbolo por tipo (item 3f). Glifos unicode em vez de imagem: a ferramenta não
// tem biblioteca de símbolos ainda (isso é fase posterior, depois das 12 etapas
// -- ver CARTOGRAFO.md), e um glifo já distingue os 6 tipos sem inventar arte
// que depois teria que ser jogada fora.
const SIMBOLO_POR_TIPO = {
  cidade: "◉",
  vila: "●",
  fortaleza: "▲",
  porto: "⚓",
  ruina: "✖",
  marco: "★",
};
// null tem o MESMO tratamento visual de "pequena" (decisão de 2026-09-23).
const TAMANHO_POR_IMPORTANCIA = { pequena: 14, media: 20, grande: 28 };

function iniciarFerramentaDeLugar(mapa, lugaresIniciais, travasIniciais) {
  const camadaLugares = L.layerGroup().addTo(mapa);
  const marcadoresPorId = {};
  let featuresAtuais = [];
  let idSelecionado = null;
  let modoAdicionar = false;
  // Cadeado geral da camada (item 1d). É um flag SEPARADO: travar a camada não
  // escreve nada em objeto nenhum, então destravar devolve cada um ao que era.
  let camadaTravada = false;

  const botaoAdicionar = document.getElementById("lugar-adicionar");
  const botaoTravaCamada = document.getElementById("trava-camada-lugares");
  const botaoDesfazer = document.getElementById("botao-desfazer");
  const botaoRefazer = document.getElementById("botao-refazer");
  const listaLateral = document.getElementById("lista-lugares");
  const campoBusca = document.getElementById("busca-lugares");

  // --- API ------------------------------------------------------------------
  async function chamar(metodo, caminho, corpo) {
    if (metodo !== "GET") mostrarSalvando();
    let resp;
    try {
      resp = await fetch(caminho, {
        method: metodo,
        headers: corpo ? { "Content-Type": "application/json" } : undefined,
        body: corpo ? JSON.stringify(corpo) : undefined,
      });
    } catch (e) {
      // Servidor fora do ar / rede: erro de gravação tem que ficar VISÍVEL.
      if (metodo !== "GET") mostrarErroDeGravacao("o servidor não respondeu");
      return { erro: "o servidor não respondeu" };
    }
    const dados = await resp.json().catch(() => ({}));
    if (!resp.ok) {
      const mensagem = dados.detail || `erro ${resp.status}`;
      // 409 = recusado por TRAVA (item 1b): aviso discreto, nunca a faixa de
      // "NÃO SALVOU" -- nada falhou, o objeto é que está protegido.
      if (resp.status === 409) {
        mostrarAviso(mensagem);
        return { erro: mensagem, travado: true };
      }
      if (metodo !== "GET") mostrarErroDeGravacao(mensagem);
      return { erro: mensagem };
    }
    if (metodo !== "GET") mostrarSalvo();
    return { dados };
  }

  // --- Trava (item 1) -------------------------------------------------------
  // A trava EFETIVA é a OU das duas: a do objeto e a da camada.
  function estaTravado(feature) {
    return camadaTravada || Boolean(feature.properties.travado);
  }
  function aplicarTravas(docTravas) {
    const camada = (docTravas.camadas || []).find((c) => c.id === "lugares");
    camadaTravada = Boolean(camada && camada.travada);
    if (botaoTravaCamada) {
      botaoTravaCamada.textContent = camadaTravada ? "🔒 camada travada" : "🔓 camada livre";
      botaoTravaCamada.classList.toggle("ativo", camadaTravada);
      botaoTravaCamada.title = camadaTravada
        ? "Destravar a camada de lugares (cada lugar volta ao estado individual dele)"
        : "Travar todos os lugares de uma vez, sem mexer no estado individual de cada um";
    }
  }
  async function recarregarTravas() {
    const r = await chamar("GET", "/api/travas");
    if (r.dados) aplicarTravas(r.dados);
  }
  if (botaoTravaCamada) {
    botaoTravaCamada.addEventListener("click", async () => {
      const r = await chamar("POST", "/api/travas/lugares", { travada: !camadaTravada });
      if (r.erro) return;
      aplicarTravas(r.dados);
      const atual = await chamar("GET", "/api/lugares");
      if (atual.dados) redesenharTudo(atual.dados);
      atualizarBotoesPilha();
    });
  }
  async function alternarTravaDe(id) {
    const feature = featuresAtuais.find((f) => f.properties.id === id);
    if (!feature) return;
    const r = await chamar("POST", `/api/lugares/${encodeURIComponent(id)}/trava`, {
      travado: !feature.properties.travado,
    });
    if (r.erro) return { erro: r.erro };
    redesenharTudo(r.dados);
    atualizarBotoesPilha();
    mostrarAviso(feature.properties.travado ? "lugar destravado" : "lugar travado");
    return {};
  }
  function alternarTravaDoSelecionado() {
    if (!idSelecionado) { mostrarAviso("nenhum lugar selecionado"); return; }
    alternarTravaDe(idSelecionado);
  }

  // --- Desenho --------------------------------------------------------------
  function iconeDoLugar(feature, selecionado) {
    const p = feature.properties;
    const tamanho = TAMANHO_POR_IMPORTANCIA[p.importancia] || TAMANHO_POR_IMPORTANCIA.pequena;
    const simbolo = SIMBOLO_POR_TIPO[p.tipo] || "●";
    const travado = estaTravado(feature);
    const classe = "icone-lugar" + (selecionado ? " selecionado" : "")
      + (p.capital ? " capital" : "") + (travado ? " travado" : "");
    // Indicação visual discreta do travado (item 1e): um cadeadinho ao lado do
    // símbolo, sem mudar o símbolo nem a cor do lugar.
    const cadeado = travado ? `<i class="cadeado-lugar">🔒</i>` : "";
    return L.divIcon({
      className: classe,
      html: `<span style="font-size:${tamanho}px;line-height:${tamanho}px">${simbolo}</span>${cadeado}`,
      iconSize: [tamanho, tamanho],
      iconAnchor: [tamanho / 2, tamanho / 2],
    });
  }

  function adicionarMarcador(feature) {
    const [lon, lat] = feature.geometry.coordinates;
    const id = feature.properties.id;
    const marcador = L.marker([lat, lon], {
      // Travado não arrasta: o jeito limpo é o marcador nem nascer arrastável
      // (item 1b), em vez de cancelar o arrasto no meio e brigar com o Leaflet.
      draggable: !estaTravado(feature),
      icon: iconeDoLugar(feature, id === idSelecionado),
      title: feature.properties.nome || id,
    });
    marcador.on("dragend", async () => {
      const p = marcador.getLatLng();
      const r = await chamar("PUT", `/api/lugares/${encodeURIComponent(id)}/posicao`, {
        lon: p.lng, lat: p.lat,
      });
      // Se o servidor recusou (ex.: caiu no mar), redesenha do estado REAL --
      // o marcador não pode ficar num lugar que o dado não tem.
      const atual = r.dados || (await chamar("GET", "/api/lugares")).dados;
      if (atual) redesenharTudo(atual);
      atualizarBotoesPilha();
    });
    marcador.on("click", (evento) => {
      L.DomEvent.stopPropagation(evento);
      selecionar(id);
      abrirModalEdicao(feature);
    });
    marcador.on("mouseover", () => travarSobCursor(
      `${feature.properties.nome || id} (${feature.properties.tipo})${estaTravado(feature) ? " 🔒 travado" : ""}`
    ));
    marcador.on("mouseout", () => destravarSobCursor());
    marcador.addTo(camadaLugares);
    marcadoresPorId[id] = marcador;
  }

  function redesenharTudo(featureCollection) {
    featuresAtuais = featureCollection.features || [];
    camadaLugares.clearLayers();
    for (const id in marcadoresPorId) delete marcadoresPorId[id];
    for (const feature of featuresAtuais) adicionarMarcador(feature);
    redesenharLista();
  }

  // --- Lista lateral com busca (item 3b) ------------------------------------
  function redesenharLista() {
    const filtro = (campoBusca.value || "").trim().toLowerCase();
    listaLateral.innerHTML = "";
    const visiveis = featuresAtuais.filter((f) => {
      if (!filtro) return true;
      const p = f.properties;
      return (p.nome || "").toLowerCase().includes(filtro) || p.id.toLowerCase().includes(filtro);
    });
    if (visiveis.length === 0) {
      const vazio = document.createElement("li");
      vazio.className = "nota";
      vazio.textContent = featuresAtuais.length === 0 ? "nenhum lugar ainda" : "nada com esse nome";
      listaLateral.appendChild(vazio);
      return;
    }
    for (const feature of visiveis) {
      const p = feature.properties;
      const item = document.createElement("li");
      item.className = "item-lugar" + (p.id === idSelecionado ? " selecionado" : "");
      item.innerHTML = `<span class="simbolo-lista">${SIMBOLO_POR_TIPO[p.tipo] || "●"}</span>`
        + `<span class="nome-lista">${p.nome || p.id}</span>`
        + `<span class="tipo-lista">${p.tipo}</span>`
        + (estaTravado(feature) ? `<span class="cadeado-lista" title="travado">🔒</span>` : "");
      item.addEventListener("click", () => {
        selecionar(p.id);
        const [lon, lat] = feature.geometry.coordinates;
        mapa.panTo([lat, lon]);
      });
      listaLateral.appendChild(item);
    }
  }
  campoBusca.addEventListener("input", redesenharLista);

  // Destaque recíproco: selecionar no mapa destaca na lista e vice-versa.
  // Só os DOIS marcadores que mudam de estado são redesenhados (o que perdeu e o
  // que ganhou o destaque), não a coleção inteira -- com muitos lugares, um
  // setIcon por marcador a cada clique seria trabalho à toa.
  function selecionar(id) {
    const anterior = idSelecionado;
    idSelecionado = id;
    // Selecionar um lugar desmarca a área selecionada, e vice-versa (areas.js):
    // a tecla T age sobre UM objeto, o último escolhido.
    if (id && typeof window.limparSelecaoDeArea === "function") window.limparSelecaoDeArea();
    for (const alvo of [anterior, id]) {
      if (!alvo) continue;
      const feature = featuresAtuais.find((f) => f.properties.id === alvo);
      const marcador = marcadoresPorId[alvo];
      if (feature && marcador) marcador.setIcon(iconeDoLugar(feature, alvo === id));
    }
    redesenharLista();
  }
  window.limparSelecaoDeLugar = () => { if (idSelecionado) selecionar(null); };

  // --- Modal (item 2) -------------------------------------------------------
  const modal = document.getElementById("modal-lugar");
  const form = document.getElementById("form-lugar");
  const tituloModal = document.getElementById("titulo-modal-lugar");
  const campoNome = document.getElementById("campo-lugar-nome");
  const campoId = document.getElementById("campo-lugar-id");
  const campoTipo = document.getElementById("campo-lugar-tipo");
  const campoImportancia = document.getElementById("campo-lugar-importancia");
  const campoCapital = document.getElementById("campo-lugar-capital");
  const campoLat = document.getElementById("campo-lugar-lat");
  const campoLon = document.getElementById("campo-lugar-lon");
  const erroModal = document.getElementById("erro-modal-lugar");
  const campoTravado = document.getElementById("campo-lugar-travado");
  const notaTravado = document.getElementById("nota-lugar-travado");
  const botaoApagar = document.getElementById("botao-lugar-apagar");
  const botaoSalvarOutro = document.getElementById("botao-lugar-salvar-outro");

  let edicaoDe = null; // id do lugar sendo editado, ou null quando é criação

  function mostrarErroNoModal(mensagem) {
    erroModal.textContent = mensagem;
    erroModal.hidden = false;
  }
  function limparErroDoModal() {
    erroModal.textContent = "";
    erroModal.hidden = true;
  }

  function sugerirId(nome) {
    return (nome || "")
      .toLowerCase()
      .normalize("NFD").replace(/[̀-ͯ]/g, "") // tira acento
      .replace(/[^a-z0-9]+/g, "-")
      .replace(/^-+|-+$/g, "");
  }
  campoNome.addEventListener("input", () => {
    // Só sugere enquanto está CRIANDO e o usuário não digitou um id próprio.
    if (edicaoDe !== null || campoId.dataset.editadoAMao === "sim") return;
    campoId.value = sugerirId(campoNome.value);
  });
  campoId.addEventListener("input", () => { campoId.dataset.editadoAMao = "sim"; });

  function atualizarNotaCapital() {
    const ehCidade = campoTipo.value === "cidade";
    campoCapital.disabled = !ehCidade;
    if (!ehCidade) campoCapital.checked = false;
    document.getElementById("nota-capital").textContent = ehCidade ? "" : "(só em cidade)";
  }
  campoTipo.addEventListener("change", atualizarNotaCapital);

  function abrirModal({ titulo, feature, lat, lon }) {
    limparErroDoModal();
    tituloModal.textContent = titulo;
    edicaoDe = feature ? feature.properties.id : null;
    const p = feature ? feature.properties : {};
    campoNome.value = p.nome || "";
    campoId.value = p.id || "";
    campoId.readOnly = Boolean(feature); // id é permanente: só editável na criação
    campoId.dataset.editadoAMao = feature ? "sim" : "";
    campoTipo.value = p.tipo || "vila";
    campoImportancia.value = p.importancia || "";
    campoCapital.checked = Boolean(p.capital);
    const coords = feature ? feature.geometry.coordinates : [lon, lat];
    campoLon.value = Number(coords[0]).toFixed(4);
    campoLat.value = Number(coords[1]).toFixed(4);
    atualizarNotaCapital();
    campoTravado.checked = Boolean(p.travado);
    atualizarEstadoTravadoNoModal(feature);
    botaoApagar.hidden = !feature;
    botaoSalvarOutro.hidden = Boolean(feature);
    modal.style.display = "flex";
    campoNome.focus();
    campoNome.select();
  }

  // Com o lugar travado, o modal vira só leitura, menos a própria trava: é o que
  // faz "destravar" ser o único caminho para editar de novo, e evita mandar pro
  // servidor uma edição que ele vai recusar. Com a CAMADA travada, nem a trava
  // individual muda (mudaria um estado sem efeito nenhum enquanto a camada
  // mandasse).
  function atualizarEstadoTravadoNoModal(feature) {
    const travadoAgora = feature ? estaTravado(feature) : camadaTravada;
    const campos = [campoNome, campoId, campoTipo, campoImportancia, campoCapital, campoLat, campoLon];
    for (const campo of campos) campo.disabled = travadoAgora;
    if (!travadoAgora) atualizarNotaCapital(); // devolve a regra "capital só em cidade"
    campoTravado.disabled = camadaTravada;
    notaTravado.textContent = camadaTravada
      ? "(a camada inteira está travada)"
      : (travadoAgora ? "(destrave para editar)" : "");
    document.getElementById("botao-lugar-salvar").disabled = travadoAgora;
    botaoApagar.disabled = travadoAgora;
    botaoSalvarOutro.disabled = travadoAgora;
  }

  campoTravado.addEventListener("change", async () => {
    if (edicaoDe === null) return; // na criação é só o valor inicial do campo
    const r = await alternarTravaDe(edicaoDe);
    const feature = featuresAtuais.find((f) => f.properties.id === edicaoDe);
    if (r && r.erro) { campoTravado.checked = !campoTravado.checked; return; }
    if (feature) atualizarEstadoTravadoNoModal(feature);
  });

  function fecharModal() {
    modal.style.display = "none";
    edicaoDe = null;
  }

  function abrirModalCriacao(latlng) {
    abrirModal({ titulo: "Novo lugar", feature: null, lat: latlng.lat, lon: latlng.lng });
  }
  function abrirModalEdicao(feature) {
    abrirModal({ titulo: `Editar "${feature.properties.nome || feature.properties.id}"`, feature });
  }

  function lerFormulario() {
    const nome = campoNome.value.trim() || null;
    const id = campoId.value.trim();
    const tipo = campoTipo.value;
    const importancia = campoImportancia.value || null;
    const capital = campoCapital.checked;
    const lat = parseFloat(campoLat.value);
    const lon = parseFloat(campoLon.value);
    return { nome, id, tipo, importancia, capital, lat, lon };
  }

  // Validação NO MODAL (nunca alert -- pedido explícito). A do servidor continua
  // valendo e é a que manda: esta aqui é só pra não fazer ida e volta por um
  // campo em branco. Erro que só o servidor sabe (id repetido, ponto no mar)
  // volta pela resposta e aparece no mesmo lugar.
  function validar(dados) {
    if (!dados.id) return "o id não pode ficar em branco";
    if (!/^[a-z0-9-]+$/.test(dados.id)) return "o id só aceita minúscula, número e hífen";
    if (!TIPOS_LUGAR.includes(dados.tipo)) return `tipo inválido: ${dados.tipo}`;
    if (dados.importancia !== null && !IMPORTANCIAS_LUGAR.includes(dados.importancia)) {
      return `importância inválida: ${dados.importancia}`;
    }
    if (dados.capital && dados.tipo !== "cidade") return "capital só vale em cidade";
    if (!Number.isFinite(dados.lat) || !Number.isFinite(dados.lon)) return "latitude/longitude precisam ser números";
    return null;
  }

  async function salvarDoModal({ criarOutro }) {
    const dados = lerFormulario();
    const erro = validar(dados);
    if (erro) { mostrarErroNoModal(erro); return; }
    limparErroDoModal();

    const propriedades = {
      tipo: dados.tipo, nome: dados.nome, capital: dados.capital, importancia: dados.importancia,
    };
    let resposta;
    if (edicaoDe === null) {
      resposta = await chamar("POST", "/api/lugares", {
        // Na criação o `travado` do formulário entra junto; depois disso ele só
        // muda pelo caminho próprio (/trava), que é o que passa pelo desfazer
        // com nome de operação legível.
        id: dados.id, lon: dados.lon, lat: dados.lat,
        propriedades: { ...propriedades, travado: campoTravado.checked },
      });
    } else {
      resposta = await chamar("PUT", `/api/lugares/${encodeURIComponent(edicaoDe)}`, { propriedades });
      if (!resposta.erro) {
        const antes = featuresAtuais.find((f) => f.properties.id === edicaoDe);
        const [lonAntes, latAntes] = antes ? antes.geometry.coordinates : [null, null];
        if (lonAntes !== dados.lon || latAntes !== dados.lat) {
          resposta = await chamar("PUT", `/api/lugares/${encodeURIComponent(edicaoDe)}/posicao`, {
            lon: dados.lon, lat: dados.lat,
          });
        }
      }
    }
    if (resposta.erro) { mostrarErroNoModal(resposta.erro); return; }

    redesenharTudo(resposta.dados);
    atualizarBotoesPilha();
    if (criarOutro) {
      // "salvar e criar outro": mantém o modal aberto, limpo, pronto pro
      // próximo -- pra marcar vários em sequência sem reabrir nada.
      const tipoAnterior = dados.tipo;
      abrirModal({ titulo: "Novo lugar", feature: null, lat: dados.lat, lon: dados.lon });
      campoTipo.value = tipoAnterior;
      atualizarNotaCapital();
      campoNome.value = "";
      campoId.value = "";
      campoId.dataset.editadoAMao = "";
      campoNome.focus();
    } else {
      fecharModal();
    }
  }

  form.addEventListener("submit", (evento) => {
    evento.preventDefault(); // Enter salva
    salvarDoModal({ criarOutro: false });
  });
  botaoSalvarOutro.addEventListener("click", () => salvarDoModal({ criarOutro: true }));
  document.getElementById("botao-lugar-cancelar").addEventListener("click", fecharModal);
  botaoApagar.addEventListener("click", async () => {
    // Sem confirmação: o desfazer cobre (item 3i).
    const id = edicaoDe;
    if (!id) return;
    const r = await chamar("DELETE", `/api/lugares/${encodeURIComponent(id)}`);
    if (r.erro) { mostrarErroNoModal(r.erro); return; }
    if (idSelecionado === id) idSelecionado = null;
    redesenharTudo(r.dados);
    atualizarBotoesPilha();
    fecharModal();
  });
  modal.addEventListener("keydown", (evento) => {
    if (evento.key === "Escape") { evento.stopPropagation(); fecharModal(); }
  });

  // --- Modo de adicionar ----------------------------------------------------
  function ativar() {
    // Camada travada não recebe lugar novo: avisar aqui é melhor do que abrir o
    // modal com todos os campos mortos e deixar o servidor recusar no fim.
    if (camadaTravada) { mostrarAviso("a camada de lugares está travada"); return; }
    modoAdicionar = true;
    botaoAdicionar.classList.add("ativo");
    mapa.getContainer().classList.add("cursor-cruz");
  }
  function desativar() {
    modoAdicionar = false;
    botaoAdicionar.classList.remove("ativo");
    mapa.getContainer().classList.remove("cursor-cruz");
  }
  botaoAdicionar.addEventListener("click", () => {
    if (modoAdicionar) desativar();
    else { desativarTodasAsFerramentas(); ativar(); }
  });
  registrarFerramenta({
    nome: "lugar", tecla: "l", ativar, desativar, estaAtiva: () => modoAdicionar,
  });

  mapa.on("click", (evento) => {
    if (!modoAdicionar || arrastandoComEspaco()) {
      if (!modoAdicionar && idSelecionado) selecionar(null); // clique no fundo desmarca
      return;
    }
    desativar();
    abrirModalCriacao(evento.latlng);
  });

  // --- Desfazer/refazer -----------------------------------------------------
  async function atualizarBotoesPilha() {
    const r = await chamar("GET", "/api/pilha");
    if (r.erro) return;
    botaoDesfazer.disabled = !r.dados.pode_desfazer;
    botaoRefazer.disabled = !r.dados.pode_refazer;
  }

  async function desfazerOuRefazer(caminho) {
    const r = await chamar("POST", caminho, {});
    if (r.erro) return;
    // A operação desfeita pode ter sido uma trava de camada (item 1f) -- as
    // travas vêm antes de redesenhar, senão o mapa nasceria com o cadeado velho.
    await recarregarTravas();
    const atual = await chamar("GET", "/api/lugares");
    if (atual.dados) redesenharTudo(atual.dados);
    // Posição de camada de referência também pode ter sido desfeita -- recarrega
    // as camadas pra imagem voltar pro lugar certo sem F5.
    if (typeof recarregarCamadasReferencia === "function") recarregarCamadasReferencia();
    // Idem para as áreas (B4): a operação desfeita pode ter sido um recorte.
    if (typeof window.recarregarAreas === "function") window.recarregarAreas();
    if (typeof window.recarregarRios === "function") window.recarregarRios();
    atualizarBotoesPilha();
  }
  botaoDesfazer.addEventListener("click", () => desfazerOuRefazer("/api/desfazer"));
  botaoRefazer.addEventListener("click", () => desfazerOuRefazer("/api/refazer"));

  aplicarTravas(travasIniciais || { camadas: [] });
  redesenharTudo(lugaresIniciais);
  atualizarBotoesPilha();
  // Os botões de desfazer/refazer valem para o log inteiro, não só para lugares:
  // a ferramenta de Área chama isto depois de cada gravação dela.
  window.atualizarBotoesDaPilha = atualizarBotoesPilha;

  // O que a interface geral (interface.js) precisa chamar: a tecla T trava/destrava
  // o lugar selecionado.
  return { alternarTravaDoSelecionado };
}

File: lore/mapas/ferramentas/backend/lugares.py
"""Ferramenta de Lugar (ponto) -- B2 da rodada noturna de 2026-09-22, esquema de
`importancia` corrigido em 2026-09-23.

Vocabulário fechado de `tipo` e as validações "capital só em cidade" / "ponto cai em
terra" vêm de `ESPEC-dados.md`. Toda gravação passa por `operacoes.registrar_operacao`
(B1) -- cada criar/mover/editar/apagar vira uma operação desfazível, gravando só a
feature afetada (correção de 2026-09-23, ver docstring de `operacoes.py`).

**Esquema de `importancia`**: `"pequena"`, `"media"`, `"grande"`, ou `null`. `null`
tem o mesmo tratamento VISUAL de `"pequena"` na ferramenta (tamanho do marcador)
-- mas o valor gravado pode ficar `null` (lugar ainda não classificado) sem que
isso force um valor no dado. Decidido na segunda rodada (2026-09-21); perdido numa
reescrita de `ESPEC-dados.md` antes do commit `96e4188` (achado em 2026-09-22,
corrigido em 2026-09-23 com o documento original recuperado pelo usuário -- ver
`ESPEC-dados.md`, "Nota de recuperação"). `capital` continua validado como
booleano (é o que faz "capital só em cidade"
funcionar).
"""

import json
from pathlib import Path

import numpy as np
from PIL import Image

from . import coordenadas, operacoes, travas

CAMADA = "lugares"  # a camada deste tipo de objeto no cadeado geral (travas.py)

RAIZ_MAPAS = Path(__file__).resolve().parents[2]
CAMINHO_LUGARES = RAIZ_MAPAS / "dados" / "lugares.geojson"
CAMINHO_COSTA = RAIZ_MAPAS / "mascaras" / "costa_10240.png"

TIPOS_VALIDOS = {"cidade", "vila", "fortaleza", "porto", "ruina", "marco"}
IMPORTANCIAS_VALIDAS = {"pequena", "media", "grande"}

_costa_array = None  # cache em memória -- uma imagem de 10240px só é aberta UMA
                      # vez por processo, nunca a cada checagem de ponto (evitaria
                      # decodificar ~104 MB repetidamente a cada lugar criado).


def _costa() -> np.ndarray:
    global _costa_array
    if _costa_array is None:
        with Image.open(CAMINHO_COSTA) as im:
            _costa_array = np.asarray(im)
    return _costa_array


def ponto_em_terra(lon: float, lat: float) -> bool:
    """255 = terra, 0 = mar (achado registrado em CARTOGRAFO.md, "Achados técnicos
    registrados") -- ponto fora da tela de 10240px também conta como 'não é terra'."""
    dados = coordenadas.carregar_coordenadas()
    projecao = dados["projecao"]
    referencia = dados["referencia"]
    px_por_grau = projecao["px_por_grau"]
    x = round(referencia["x_meridiano_zero_px"] + lon * px_por_grau)
    y = round(referencia["y_equador_px"] - lat * px_por_grau)
    arr = _costa()
    if not (0 <= y < arr.shape[0] and 0 <= x < arr.shape[1]):
        return False
    return bool(arr[y, x] >= 128)


def carregar() -> dict:
    with open(CAMINHO_LUGARES, encoding="utf-8") as f:
        return json.load(f)


def validar_propriedades(propriedades: dict) -> None:
    tipo = propriedades.get("tipo")
    if tipo not in TIPOS_VALIDOS:
        raise ValueError(f"'tipo' tem que ser um de {sorted(TIPOS_VALIDOS)}, recebi {tipo!r}")
    capital = propriedades.get("capital", False)
    if not isinstance(capital, bool):
        raise ValueError("'capital' tem que ser booleano")
    if capital and tipo != "cidade":
        raise ValueError("'capital' só pode ser True quando 'tipo' é 'cidade'")
    importancia = propriedades.get("importancia")
    if importancia is not None and importancia not in IMPORTANCIAS_VALIDAS:
        raise ValueError(
            f"'importancia' tem que ser null ou um de {sorted(IMPORTANCIAS_VALIDAS)}, recebi {importancia!r}"
        )
    travado = propriedades.get("travado", False)
    if not isinstance(travado, bool):
        raise ValueError("'travado' tem que ser booleano")
    # Visibilidade para o jogador (B3, 2026-09-23 noite): ausente vale true. False
    # tira o lugar (e o nome dele) da versão do jogador de toda exportação.
    if not isinstance(propriedades.get("visivel_jogador", True), bool):
        raise ValueError("'visivel_jogador' tem que ser booleano")


def criar_lugar(id_lugar: str, lon: float, lat: float, propriedades: dict) -> dict:
    # Camada travada é INERTE: não recebe objeto novo também. Decisão da IA (o
    # pedido fala de mover e apagar); é o que um cadeado de camada significa em
    # qualquer editor, e destravar desfaz a restrição sem deixar rastro.
    travas.exigir_camada_livre(CAMADA, "criar um lugar")
    validar_propriedades(propriedades)
    if not ponto_em_terra(lon, lat):
        raise ValueError("o ponto cai fora de terra (mar, ou fora da tela) -- lugar tem que ficar em terra")

    dados = carregar()
    if any(f["properties"]["id"] == id_lugar for f in dados["features"]):
        raise ValueError(f"id já existe: {id_lugar}")

    feature = {
        "type": "Feature",
        "geometry": {"type": "Point", "coordinates": [lon, lat]},
        # `travado` sempre presente no dado (padrão false, ESPEC-dados.md): um
        # campo ausente e um `false` valem o mesmo na leitura, mas gravar
        # explícito deixa o arquivo legível sem consultar o padrão.
        "properties": {"id": id_lugar, **propriedades, "travado": bool(propriedades.get("travado", False))},
    }
    operacoes.registrar_operacao(
        "criar_lugar", "dados/lugares.geojson",
        {id_lugar: {"antes": None, "depois": feature}},
    )
    return feature


def _achar(dados: dict, id_lugar: str) -> dict:
    for f in dados["features"]:
        if f["properties"]["id"] == id_lugar:
            return f
    raise KeyError(f"lugar desconhecido: {id_lugar}")


def mover_lugar(id_lugar: str, lon: float, lat: float) -> dict:
    dados = carregar()
    feature_antes = _achar(dados, id_lugar)
    # A trava é checada ANTES da validação de terra: um lugar travado nem chega a
    # ser avaliado, e o aviso que o usuário recebe é "está travado", não "caiu no
    # mar" (seriam dois motivos diferentes para a mesma recusa).
    travas.exigir_objeto_livre(CAMADA, feature_antes["properties"], "mover este lugar")
    if not ponto_em_terra(lon, lat):
        raise ValueError("o ponto cai fora de terra (mar, ou fora da tela) -- lugar tem que ficar em terra")
    feature_depois = json.loads(json.dumps(feature_antes))
    feature_depois["geometry"]["coordinates"] = [lon, lat]
    operacoes.registrar_operacao(
        "mover_lugar", "dados/lugares.geojson",
        {id_lugar: {"antes": feature_antes, "depois": feature_depois}},
    )
    return feature_depois


def editar_lugar(id_lugar: str, propriedades: dict) -> dict:
    dados = carregar()
    feature_antes = _achar(dados, id_lugar)
    travas.exigir_objeto_livre(CAMADA, feature_antes["properties"], "editar este lugar")
    novas_propriedades = {**feature_antes["properties"], **propriedades, "id": id_lugar}
    validar_propriedades(novas_propriedades)
    feature_depois = json.loads(json.dumps(feature_antes))
    feature_depois["properties"] = novas_propriedades
    operacoes.registrar_operacao(
        "editar_lugar", "dados/lugares.geojson",
        {id_lugar: {"antes": feature_antes, "depois": feature_depois}},
    )
    return feature_depois


def apagar_lugar(id_lugar: str) -> None:
    dados = carregar()
    feature_antes = _achar(dados, id_lugar)  # KeyError se não existir
    travas.exigir_objeto_livre(CAMADA, feature_antes["properties"], "apagar este lugar")
    operacoes.registrar_operacao(
        "apagar_lugar", "dados/lugares.geojson",
        {id_lugar: {"antes": feature_antes, "depois": None}},
    )


def definir_trava(id_lugar: str, travado: bool) -> dict:
    """Trava/destrava UM lugar. Caminho próprio, e não `editar_lugar`, por dois
    motivos: `editar_lugar` recusa objeto travado (senão nada travado poderia ser
    destravado), e a operação no log precisa se chamar "travar_lugar" pra ficar
    legível. A trava DE CAMADA continua mandando: com a camada travada, nem o
    estado individual muda (seria uma mudança sem efeito nenhum, e confusa)."""
    if not isinstance(travado, bool):
        raise ValueError("'travado' tem que ser booleano")
    travas.exigir_camada_livre(CAMADA, "mudar a trava deste lugar")
    dados = carregar()
    feature_antes = _achar(dados, id_lugar)
    feature_depois = json.loads(json.dumps(feature_antes))
    feature_depois["properties"]["travado"] = travado
    operacoes.registrar_operacao(
        "travar_lugar" if travado else "destravar_lugar", "dados/lugares.geojson",
        {id_lugar: {"antes": feature_antes, "depois": feature_depois}},
    )
    return feature_depois
